tests/decomposition/test_decomposition_graph.py

```diff
@@ -0,0 +1,181 @@
+        # 5 ops and 2 decompositions (RY is in the target gate set now)
+        assert len(graph2._graph.nodes()) == 7
+        # 6 edges from ops to decompositions and 2 from decompositions to ops
+        assert len(graph2._graph.edges()) == 8
+
+    @pytest.mark.unit
+    def test_graph_solve(self, _):
+        """Tests solving a simple graph for the optimal decompositions."""
+
+        op = qml.Hadamard(wires=[0])
+        graph = DecompositionGraph(
+            operations=[op],
+            target_gate_set={"RX", "RY", "RZ", "GlobalPhase"},
+        )
+        graph.solve()
+
+        # verify that the better decomposition rule is chosen when both are valid.
+        expected_resource = Resources(
+            {
+                qml.resource_rep(qml.RZ): 1,
+                qml.resource_rep(qml.RY): 1,
+                qml.resource_rep(qml.GlobalPhase): 1,
+            },
+        )
+        assert graph.resource_estimate(op) == expected_resource
+        assert graph.decomposition(op).compute_resources() == expected_resource
+
+    @pytest.mark.unit
+    def test_decomposition_not_found(self, _):
+        """Tests that the correct error is raised if a decomposition isn't found."""
+
+        op = qml.Hadamard(wires=[0])
+        graph = DecompositionGraph(operations=[op], target_gate_set={"RX", "RY", "GlobalPhase"})
+        with pytest.raises(DecompositionError, match="Decomposition not found for {'Hadamard'}"):
+            graph.solve()
+
+    @pytest.mark.unit
+    def test_decomposition_with_resource_params(self, _):
+        """Tests operators with non-empty resource params."""
+
+        class CustomOp(qml.operation.Operation):  # pylint: disable=too-few-public-methods
+            """A custom operation."""
+
+            resource_keys = {"num_wires"}
+
+            @property
+            def resource_params(self):
+                return {"num_wires": len(self.wires)}
+
+        def _custom_resource(num_wires):
+            return {
+                qml.resource_rep(qml.MultiRZ, num_wires=num_wires): 1,
+                qml.resource_rep(qml.MultiRZ, num_wires=num_wires - 1): 2,
+            }
+
+        @qml.register_resources(_custom_resource)
+        def _custom_decomp(*_, **__):
+            raise NotImplementedError
+
+        decompositions[CustomOp] = [_custom_decomp]
+
+        op = CustomOp(wires=[0, 1, 2, 3])
+        graph = DecompositionGraph(
+            operations=[op],
+            target_gate_set={"RX", "RZ", "CZ", "GlobalPhase"},
+        )
+        # 10 ops and 7 decompositions (1 for the custom op, 1 for each of the two MultiRZs,
+        # 1 for CNOT, 2 for Hadamard, and 1 for RY)
+        assert len(graph._graph.nodes()) == 17
+        # 16 edges from ops to decompositions and 7 from decompositions to ops
+        assert len(graph._graph.edges()) == 23
+
+        graph.solve()
+        assert graph.resource_estimate(op) == Resources(
```

Comment:
I don't know how much we expect the test suite to grow, but maybe substantially. 
As we already have a conftest.py file, it might be nice to provide a simple wrapper that takes a `R: dict[class, int]` and maps it to `Resources({qml.resource_rep(key): val for key, val in R.items()})` :shrug:

Reply:
So test writing people would not have to think about conversion to resources but could do `to_resources({qml.CZ: 10, qml.RX: 4})`, I mean :)